Extract content selection into separate component

robertknight opened the pull request:
Extract the logic for presenting a list of available content sources and showing the dialogs for choosing content from specific sources from `FilePickerApp` into a separate `ContentSelector` component. This will keep FilePickerApp more manageable as new steps are added to the assignment configuration flow.

There should be no functional or visible changes here.

Part of https://github.com/hypothesis/lms/issues/2524.

esanzgar reviewed (approved):
I like the refactoring into different components. Thanks!

I left a few comments for your consideration.

robertknight commented:
https://github.com/hypothesis/lms/pull/2646 is now almost ready. The changes from that will need to be incorporated into this PR when it lands.

robertknight commented on lms/static/scripts/frontend_apps/components/FilePickerApp.js:
```diff
@@ -83,87 +62,12 @@ export default function FilePickerApp({
    * Flag indicating whether the form should be auto-submitted on the next
    * render.
    */
   const [shouldSubmit, submit] = useState(false);
```
You're right that this flag is currently not needed. However it will be needed again soon when introducing a new step to the assignment configuration flow to choose the groups configuration.

robertknight commented on lms/static/scripts/frontend_apps/components/ContentSelector.js:
```diff
@@ -0,0 +1,186 @@
+ * @prop {string} title
+ * @prop {Error} error
+ *
+ * @typedef {'lms'|'url'|null} DialogType
```
`null` is a special case here though that means "no dialog is shown". I understand that using a string would make sense for the object-switch pattern you suggested below, but using a falsey value means you can use things like `activeDialog && {render something if any dialog is shown}`.

robertknight commented on lms/static/scripts/frontend_apps/components/ContentSelector.js:
```diff
@@ -0,0 +1,186 @@
+ * @typedef ContentSelectorProps
+ * @prop {(ei: ErrorInfo) => void} onError
+ * @prop {(c: Content) => void} onSelectContent
+ * @prop {DialogType} [defaultActiveDialog]
```
👍 - Done

robertknight commented on lms/static/scripts/frontend_apps/components/ContentSelector.js:
```diff
@@ -0,0 +1,186 @@
+      break;
+    default:
+      dialog = null;
+  }
```
While this is a little more compact, I prefer using a falsey value to represent "no dialog shown" because you can treat the value as a boolean in cases where you want to test "is this value present"? (eg. if you wanted to render something whenever any or no dialog is shown).

robertknight commented on lms/static/scripts/frontend_apps/components/ContentSelector.js:
```diff
@@ -0,0 +1,186 @@
+ * @param {ContentSelectorProps} props
+ */
+export default function ContentSelector({
+  defaultActiveDialog = null,
+  onError,
+  onSelectContent,
```
I have re-ordered the typedef above so that the props are listed in alphabetical order and consistent with the definition here.

robertknight commented on lms/static/scripts/frontend_apps/components/test/ContentSelector-test.js:
```diff
@@ -0,0 +1,333 @@
+      enablePublicViewing: sinon.stub(),
+    });
+
+    fakeConfig = {
```
I'd be careful here, since different components use different subsets of the configuration. Within a given context (eg. the file picker app) the same fake configuration could be used. For different apps (file picker, LTI launch etc.) then different fakes will be needed though.

robertknight commented:
I have rebased this on master, adapting to the recent UI changes to the buttons in the content source list. I also made a few minor corrections based on feedback and re-reviewing the code:

- The `"lms"` enum value for `DialogType` was renamed to `lmsFile`
- The ordering of props in the `ContentSelectorProps` typedef now matches the ordering of props in the `ContentSelector` function
- I renamed several of the `data-testid` values to be slightly clearer
- I moved the CSS classes for components rendered by `ContentSelector` into a `ContentSelector.scss` SASS module and renamed them accordingly

There should be no functional changes here.

I'm going to go ahead and merge this on the basis of @esanzgar's previous review.

robertknight commented on lms/static/styles/components/_ContentSelector.scss:
```diff
@@ -0,0 +1,38 @@
+@use "../variables" as var;
+
+.ContentSelector__loading-backdrop {
```
These styles have been moved from the `FilePickerApp.scss` file with no changes other than changing the component name from `FilePickerApp` to `ContentSelector`.
